Add IScheduleObject

Simn opened the pull request:
As my war against closures continues, this adds a new function to the scheduler that allows scheduling `IScheduleObject` instances without delay. We have a lot of small objects that schedule a closure to be called asynchronously, in which case it makes sense to instead schedule the objects themselves to avoid the closure allocation.

This now comes with the added benefit of avoiding the Int64 allocation on targets that don't natively support it. Numbers look ok:

100k | kt_coro | IScheduleObject | scheduler_state | +suspendCancellable
--- | --- | --- | --- | ---
neko | 107.389 | 101.435 | 115.645 | 25.528
python | 34.672 | 28.449 | 36.087 | 20.750
eval | 20.219 | 17.761 | 21.360 | 13.401
hl | 10.020 | 10.041 | 11.006 | 7.427
cpp | 2.769 | 2.788 | 2.838 | 2.242
js | 2.091 | 1.975 | 2.079 | 1.907
jvm | 1.424 | 1.401 | 1.411 | 1.305

HL is a disappointment here, I guess that the overhead from the interface call offsets the reduced GC activity. But this isn't so much about run-time anyway as it is about memory footprint:

<img width="1771" alt="jprofiler_9f93aEHnFs" src="https://github.com/user-attachments/assets/fa7a5c8e-cede-4b3b-b507-fcacde39fa42" />

We traded 8M continuation closures + 2M async closures for 2M `FunctionScheduleObject`, which is the wrapper type for where we're still calling `schedule(0, ...)`. It should be possible to reduce this further by identifying these places. The new `resumeAsync` convenience functions and its siblings might actually get in the way of that a little.

Memory consumption goes from 1.95GB with 5.7M allocated objects to 1.77GB with 4.3M allocated objects.

Aidan63 commented:
Are those numbers derrived from the benchmark you added this morning? I've made some more changes on the state branch and these are the numbers I get for jvm.

ScheduleObject

![image](https://github.com/user-attachments/assets/fb49c41d-02c6-41f7-a1e0-b7c1d14b149f)

SchedulerState

![image](https://github.com/user-attachments/assets/6f85e393-94f8-40c1-94ec-68f720a4462e)

Are the numbers ms execution or iterations done in a given time span (Guessing something like the latter since it decreases)?

Simn commented:
Ah sorry for the confusion, the numbers I usually post are from the initial example at #121, which I've just updated after a minor modification. This measures execution time, so lower is better.

The benchmarks in tests/bench measure how often something executes within a given time frame, so higher is better for those.

Simn commented:
I've adapted `CancellingContinuation` as well.

The allocation numbers I posted initially were pretty volatile, so I let it record every single allocation (which takes much longer) to get more accurate numbers:

1M | mem | #objects
--- | --- | ---
kt_coro | 1.86GB | 60.1M
IScheduleObject | 1.71GB | 49.6M
with CancellingContinuation | 1.72GB | 48.3M
with suspendCancellable | 1.51GB | 44.7M

Memory consumption on the latest change indeed seems to increase by a small amount, which likely comes from the additional `result` and `error` fields I had to add to `CancellingContinuation`. But we allocate 1.3M less objects, so that should be worth it. 

Python goes to 26.318 from ~28.5 before, eval to 17.078 from 17.6. HL seems to go from ~10 to ~9 which I find a bit hard to believe, but HL performance is a general mystery to me. Other targets don't hate it either.

Simn commented:
I'm fixing #138 here too because I'm making changes to `CancellingContinuation` anyway. All targets love this, I've updated the table (yes the neko value is real, I double checked that).

It doesn't have a `resolve` method like `RacingContinuation`. Does it need one? I'm not sure how that would interact with the other two "paths" (resume and cancel) and the tests work fine without it.

Edit: Updated the memory consumption table as well. The C++ target _still_ doesn't survive the 1M version unfortunately!

Simn commented:
Since this has somehow turned into somewhat of a general purpose performance PR, I've adapted my benchmark to use various things like semaphores, atomics and yield:

```haxe
import hxcoro.concurrent.AtomicInt;
import hxcoro.concurrent.CoroSemaphore;
import hxcoro.ds.Channel;
import haxe.Timer;
import hxcoro.CoroRun;
import hxcoro.Coro.*;

function main() {
	final startTime = Timer.stamp();
	// half will complete, half will cancel
	final numTasks = 20000;
	var aggregateValue = 0;
	final stats = CoroRun.runScoped(node -> {
		// can become a condition variable once we have them
		final lockChannel = new Channel();
		final channel = new Channel();
		final numTasksHalved = numTasks >> 1;
		// lets only half the tasks through
		final semaphore = new CoroSemaphore(numTasksHalved);
		var completedWriters = 0;
		// used to detect when to unlock lockChannel
		var completedReaders = new AtomicInt(0);
		for (_ in 0...numTasks) {
			node.async(_ -> {
				delay(Std.random(1000));
				channel.write(1);
				yield();
				completedWriters++;
			});
		}
		for (_ in 0...numTasks) {
			node.async(_ -> {
				delay(Std.random(1000));
				semaphore.acquire();
				aggregateValue += channel.read();
				// cause more chaos by yielding
				yield();
				if (completedReaders.add(1) == numTasksHalved - 1) {
					// unlock lock channel
					lockChannel.write(0);
				}
			});
		}
		// wait for lockChannel to be released at half task time
		lockChannel.read();
		// cancel any remaining children
		node.cancelChildren();
		node.awaitChildren();
		return {
			completedWriters: completedWriters,
			completedReaders: completedReaders.load()
		};
	});
	final diff = Std.string(Timer.stamp() - startTime);
	final dot = diff.indexOf(".");
	final diff = diff.substr(0, dot + 4);
	trace('$diff seconds, value $aggregateValue, ${stats.completedReaders} completed readers, ${stats.completedWriters} completed writers');
}
```

Numbers still look good:

20k | kt_coro | IScheduleObject
--- | --- | ---
python | 14.960 | 12.520
neko | 9.995 | 5.906
eval | 6.099 | 4.800
hl | 3.265 | 2.731
js | 2.235 | 2.430
cpp | 1.527 | 1.330
jvm | 1.286 | 1.263


40k | kt_coro | IScheduleObject | cancellationException
--- | --- | --- | ---
python | 30.960 | 25.473 | 10.781
neko | 26.542 | 12.100 | 11.763
eval | 11.949 | 9.567 | 7.073
hl | 6.999 | 5.335 | 3.903
js | 3.392 | 3.779 | 1.535
cpp | 2.082 | 1.666 | 1.670
jvm | 1.427 | 1.440 | 1.212

This fixes the scaling of neko and I've confirmed that this holds up at 80k (24s) and 160k (50s). All other targets also still scale (sub-)linearly.

JS has worse numbers, which I think comes from `RacingContinuation` now using `AtomicInt` instead of the noop stub mutex, which means it's actually synchronizing something.

JVM doesn't really care one way or another, but that runtime is something else anyway... At 640k C++ goes from crashing to completing in 15s, which is slightly better.

Simn commented:
I already found a problem with that benchmark (we allocate non-trivial amounts of `CancellationException`), but that's a story for another PR. This one should be good to go.
